powerfulseal/policy/action_nodes_pods.py

```diff
@@ -0,0 +1,235 @@
+            return False
+        attr = criterion.get("name")
+        value = getattr(candidate, attr)
+        expr = re.compile(criterion.get("value"))
+        if type(value) is list:
+            return any([
+                expr.match(str(v))
+                for v in value
+            ])
+        else:
+            value = str(value)
+        return expr.match(value)
+
+    def filter(self, items):
+        """ Applies various filters based on the given policy.
+        """
+        filters = self.schema.get("filters", [])
+        mapping = {
+            "property": self.filter_property,
+            "dayTime": self.filter_day_time,
+            "randomSample": self.filter_random_sample,
+            "probability": self.filter_probability,
+        }
+        return self.filter_mapping(items, filters, mapping)
+
+    def filter_property(self, candidates, criterion):
+        """ Filters out things which don't match their property filters.
+        """
+        return [
+            candidate for candidate in candidates
+            if self.match_property(candidate, criterion)
+        ]
+
+    def filter_day_time(self, candidates, criterion, now=None):
+        """ Passed unchanged list of candidates, if the execution time
+            satisfies the policy requirements.
+        """
+        now = now or datetime.now()
+        self.logger.info("Now is %r", now)
+
+        # check the day is permitted
+        day_name = calendar.day_name[now.weekday()].lower()
+        permitted_days = criterion.get("onlyDays", [])
+        if permitted_days and day_name not in permitted_days:
+            self.logger.info("Not allowed on %s", day_name)
+            return []
+
+        # check the time is not too early
+        start = criterion.get("startTime", {})
+        start_date = now.replace(
+            hour=start.get("hour", 10),
+            minute=start.get("minute", 0),
+            second=start.get("second", 0),
+        )
+        if now < start_date:
+            self.logger.info("Too early")
+            return []
+
+        # check the time is not too late
+        end = criterion.get("endTime", {})
+        end_date = now.replace(
+            hour=end.get("hour", 15),
+            minute=end.get("minute", 59),
+            second=end.get("second", 59),
+        )
+        if now > end_date:
+            self.logger.info("Too late")
+            return []
+
+        return candidates
+
+    def filter_random_sample(self, candidates, criterion):
+        """ Returns a random sample from the initial list.
+            It supports policy `size` and `ratio` features.
+        """
+        if not criterion:
+            return []
+        size = criterion.get("size")
+        if size is None:
+            ratio = criterion.get("ratio", 1)
+            size = int(len(candidates)*ratio)
+        if size == 0:
+            self.logger.info("RandomSample size 0")
+            return []
+        return random.sample(candidates, size)
```

Comment:
If the sample size is smaller than candidate's size, a ValueError is raised [[0](https://docs.python.org/3/library/random.html#random.sample)]

Do we check it somewhere?

Reply:
Good find. Thanks!